lib/lpc55-rot-startup/src/images.rs

```diff
@@ -56,20 +42,85 @@ extern "C" {
 const PAGE_SIZE: u32 = BYTES_PER_FLASH_PAGE as u32;
 
 pub struct Image {
+    // The boundaries of the flash slot.
     flash: Range<u32>,
-    vector: &'static ImageVectors,
+    // The initial span of programmed flash pages.
+    programmed: Range<u32>,
+    // The number of pages that contain no erased flash words.
+    // _count: usize,
+    // Measurement over the `count`ed pages.
+    fwid: [u8; 32],
 }
 
-pub fn image_details(img: Image, flash: &mut Flash) -> RotImageDetails {
+impl Image {
+    // Before doing any other work with a chunk of flash memory:
+    //
+    //   - Define the address boundaries.
+    //   - Determine the bounds of the initial programmed extent.
+    //   - Count the total number of fully programmed pages.
+    //   - Measure all programmed pages including those outside of the
+    //     initial programmed extent.
+    //
+    // Note: if partially programmed pages is a possibility then that could be
+    // a problem with respect to catching exfiltration attempts.
+    pub fn new(dev: &mut Flash, flash: Range<u32>) -> Image {
+        // let mut _count = 0;
+        let mut end: Option<u32> = None;
+        let mut hash = Sha3_256::new();
+        for start in flash.clone().step_by(BYTES_PER_FLASH_PAGE) {
+            if dev.is_page_range_programmed(start, PAGE_SIZE) {
+                // _count += 1;
+                let page = unsafe {
+                    core::slice::from_raw_parts(
+                        start as *const u8,
+                        BYTES_PER_FLASH_PAGE,
+                    )
+                };
+                hash.update(page);
+            } else if end.is_none() {
+                end = Some(start);
+            }
+        }
+        let fwid = hash.finalize().try_into().unwrap_lite();
+        let programmed = Range {
+            start: flash.start,
+            end: end.unwrap_or(flash.start),
+        };
+        Image {
+            flash,
+            programmed,
+            // _count,
+            fwid,
+        }
+    }
+
+    pub fn is_programmed(&self, addr: &u32) -> bool {
+        return self.programmed.contains(addr);
+    }
+
+    pub fn is_span_programmed(&self, start: u32, length: u32) -> bool {
+        if let Some(end) = start.checked_add(length) {
+            if !self.is_programmed(&start) || !self.is_programmed(&end) {
+                false
+            } else {
+                true
+            }
+        } else {
+            false
+        }
+    }
+}
```

Comment:
I'm wary of this. It assumes we've really checked the programmed range and it assumes no gaps in programming.

Reply:
We checked the programmed range in Image::new() and you can't call this without having obtained an Image struct through Image::new()

Reply:
Having chatted about this yesterday I understand better where this is coming from and why it works.